File: railm/lib/models/status.dart
// SPDX-License-Identifier: GPL-2.0
// Author: xunicatt
// Project: railm (railm) 
// Copyright (c) 2026 xunicatt <contact.aniket.biswas@gmail.com>

import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:railm/configs/configs.dart';

class Status {
    final String number;
    final String station;
    final String time;

    const Status({
        required this.number,
        required this.station,
        required this.time,
    });

    factory Status.fromJson(Map<String, dynamic> json) {
        return Status(
            number: json["number"],
            station: json["station"],
            time: json["time"],
        );
    }

    static Future<Status?> fetchStatus(String number) async {
        final String url = "${Configs.railApiBaseUrl}/status/$number?auth=${Configs.railApiToken}";
        var response = await http.get(Uri.parse(url));
        var data = jsonDecode(response.body) as Map<String, dynamic>;

        if ((data.containsKey("status") && data["status"] == "failed") ||
                (data.containsKey("reason") && data["reason"] == "unknown")) {
            return null; 
        }

        return Status.fromJson(data);
    }

    static Future<bool> updateStatus(String trainNumber, String stationId, String time) async {
        final String url = "${Configs.railApiBaseUrl}/status/$trainNumber/$stationId?auth=${Configs.railApiToken}&time=$time";
        var response = await http.post(Uri.parse(url));
        var data = jsonDecode(response.body) as Map<String, dynamic>;

        return data.containsKey("status") && data["status"] == "success";
    }
}

File: railapi/api/status.go
// SPDX-License-Identifier: GPL-2.0
// Author: xunicatt
// Project: railm (railapi) 
// Copyright (c) 2026 xunicatt <contact.aniket.biswas@gmail.com>

package api

import (
	"encoding/json"
	"log"
	"net/http"
	"railapi/internals/db"
	"railapi/internals/models"
	"railapi/internals/rank"
)

func (c *Context) GetStatus(w http.ResponseWriter, r *http.Request) {
	defer r.Body.Close()

	number := r.PathValue("number")
	status, err := db.GetStatus(c.sql, number)
	if err != nil {
		log.Printf(
			"failed to get train 'status': %v",
			err.Error(),
		)
		serverError(w)
		return
	}

	if status == nil {
		failed(w, "unknown")
		return
	}

	data, err := json.Marshal(status)
	if err != nil {
		log.Printf(
			"failed to encode 'status': %v",
			err.Error(),
		)
		serverError(w)
		return
	}

	w.Write(data)
}

func (c *Context) UpdateStatus(w http.ResponseWriter, r *http.Request) {
	defer r.Body.Close()

	number := r.PathValue("number")
	station := r.PathValue("station")

	time := r.URL.Query().Get("time")
	if time == "" {
		failed(w, "This api is deprecated. Please check new api-docs.txt.")
		return
	}

	if rank.Check(number, station) {
		status, err := db.GetStatus(c.sql, number)
		if err != nil {
			log.Printf(
				"failed to get train status: %v",
				err.Error(),
			)
			serverError(w)
			return
		}

		status2 := &models.Status{
			Number: number,	
			Station: station,
			Time: time,
		}

		if status == nil {
			err = db.InsertStatus(c.sql, status2)
			if err != nil {
				log.Printf(
					"failed to get train status: %v",
					err.Error(),
				)
				serverError(w)
				return
			}

			success(w)
			return
		}

		if status.Station == status2.Station &&
		   status.Time == status2.Time {
			success(w)
			return
		}

		err = db.UpdateStatus(c.sql, status2)
		if err != nil {
			log.Printf(
				"failed to update train status: %v",
				err.Error(),
			)
			serverError(w)
			return
		}
	}

	success(w)
}
